### PR TITLE
Step to validate airflow dags as part of CI

### LINKED ISSUE #1443: Airflow DAG tag validation is missing
# Airflow DAG tag validation is missing

Our DAGs now include tags, however, we need a way to validate that all DAGs are tagged correctly. This means:
1. They contain all required tags
2. Tags specified have correct values.

We believe this check should occur in CircleCi before merging is allowed.

### PR DESCRIPTION
**Pull request by kik-kik**
# Step to validate airflow dags as part of CI

## Description
- `bin/test-dag-tags.py` script introduced to query `dag_tag` Airflow table to retrieve a list of all DAGs along with their tags and ensure they are correct.
- `test-environment` check now includes a command for executing a python script validating Airflow DAGs (part of `test-parse` script).
- installing mysql-client on airflow image to be able to query the DB for tag validation
- Added tag assignment to DAGs that were missing them

## Next step
- Integrate this check with `bigquery-etl` repos

resolves #1443

### CODE REVIEW

**Review by haroldwoo — APPROVED** [review]

File: bin/test-dag-tags.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+REQUIRED_TAG_TYPES = ("impact",)
+
+
+class DagValidationError(Exception):
+    pass
+
+
+def retrieve_existing_airflow_dags():
+    output = subprocess.run(
+        "docker-compose exec web mysql -Ns -h db -u root -psecret airflow -e 'SELECT dag_id, name FROM dag_tag;'",
```
Comment (haroldwoo) [inline bin/test-dag-tags.py:16]: Worth exploring if the airflow command line is able to get you tag information instead.

Comment (kik-kik) [reply bin/test-dag-tags.py:16]: I checked, unfortunately it is not possible.

File: Dockerfile [hunk 2]
```diff
@@ -17,7 +17,7 @@ RUN apt-get update && \
     apt-get install -y --no-install-recommends \
         apt-transport-https build-essential curl git libpq-dev python-dev \
         default-libmysqlclient-dev gettext sqlite3 libffi-dev libsasl2-dev \
-        lsb-release gnupg vim screen procps && \
+        lsb-release gnupg vim screen procps default-mysql-client && \
```
[the remark this answers is not in the document]
Comment (kik-kik) [reply Dockerfile:20]: that change was already there. I'm open to suggestions.

File: bin/test-dag-tags.py (hunk 1 above)
Comment (haroldwoo) [reply bin/test-dag-tags.py:16]: I would parse the python files on the filesystem rather than relying on a mysql table. The table may change over time, and also dag state for deleted dags, etc. is not always clear. Also we have multiple github repos

Comment (haroldwoo) [reply bin/test-dag-tags.py:16]: Does the circleCi environment even have the dags loaded into mysql?

Comment (kik-kik) [reply bin/test-dag-tags.py:16]: yes, it happens automatically when airflow spins up. I assume that's actually how airflow command for listing dags works. It most likely fetches them from the `dags` table.

Anyways, I added an alternative approach to checking dag tags.

Comment (kik-kik) [reply bin/test-dag-tags.py:16]: Please let me know what you think and then I'll make final tweaks.

File: bin/test-dag-tags.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        "Tag.ImpactTier.tier_2": "impact/tier_2",
+        "Tag.ImpactTier.tier_3": "impact/tier_3",
+        "repo/telemetry-airflow": "impact/tier_3",
+        "repo/bigquery-etl": "repo/bigquery-etl",
+    }
+
+
+    dags = defaultdict(list)
+
+    for file in glob.glob("dag_dir/*.py"):
```
Comment (haroldwoo) [inline bin/test-dag-tags.py:49]: does .glob do recursive lookups

Comment (haroldwoo) [reply bin/test-dag-tags.py:49]: also this may pull in library files that we've written

File: bin/test-parse [hunk 4]
```diff
@@ -88,6 +88,15 @@ function main() {
         echo "Unit tests passed!"
     fi
 
+    # Validate Airflow tags
+    # see Github Issue: https://github.com/mozilla/telemetry-airflow/issues/1443
+    python3 bin/test-dag-tags.py
```
Comment (haroldwoo) [inline bin/test-parse:93]: i would put this in a separate script, and a separate CircleCI test check as well

File: bin/test-dag-tags.py (hunk 1 above)
Comment (haroldwoo) [reply bin/test-dag-tags.py:16]: added comments. I still prefer the non sql approach. we can let klukas chime in also

Comment (haroldwoo) [reply bin/test-dag-tags.py:16]: after some thought, i dont feel too strongly about this. If it breaks we can just repair

File: bin/test-parse (hunk 4 above)
Comment (kik-kik) [reply bin/test-parse:93]: yes, if we decide to go with the parsing approach then I'd move it into it's own step in circleci

File: bin/test-dag-tags.py (hunk 3 above)
Comment (kik-kik) [reply bin/test-dag-tags.py:49]: I tried it locally and it seems to behave as expected. I can double check to make sure thought.

**Comment by kik-kik** [comment]
@jklukas what are your thoughts?

File: bin/test-dag-tags.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    for dag in output.replace("\r", "").split("\n"):
+        dag_name, tag_name = dag.split("\t")
+
+        dags[dag_name] = dags[dag_name] + [tag_name]
+
+    return dags
+
+
+import glob, os
+import re
```
Comment (jklukas) [inline bin/test-dag-tags.py:32]: Can these imports be at the top of the file?

File: bin/test-dag-tags.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+# for file in glob.glob("*.txt"):
+#     print(file)
```
Comment (jklukas) [inline bin/test-dag-tags.py:35]: Should remove this before merging.

File: bin/test-dag-tags.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+def get_dag_tags(dag_dir: str):
+    tag_mapping = {
+        "Tag.ImpactTier.tier_1": "impact/tier_1",
+        "Tag.ImpactTier.tier_2": "impact/tier_2",
+        "Tag.ImpactTier.tier_3": "impact/tier_3",
+        "repo/telemetry-airflow": "impact/tier_3",
+        "repo/bigquery-etl": "repo/bigquery-etl",
+    }
```
Comment (jklukas) [inline bin/test-dag-tags.py:44]: Maintaining this mapping here seems problematic. I think this should be generated from the enum rather than duplicated here.

File: bin/test-dag-tags.py (hunk 1 above)
Comment (jklukas) [reply bin/test-dag-tags.py:16]: I prefer the sql approach, as it involves working with parsed DAGs. The textual approach could have a lot of unexpected edge cases and will be brittle based on having to maintain mappings from enum constructs to text values.

Comment (haroldwoo) [reply bin/test-dag-tags.py:16]: The sql approach is more correct. The parsing would be quicker, and less brittle imo. I don't feel strongly either way, and will rely on kik to be the go to person for ci issues regarding this in the future

Comment (kik-kik) [reply bin/test-dag-tags.py:16]: 🙈 <- literally me right now 😅

Comment (haroldwoo) [reply bin/test-dag-tags.py:16]: if you go sql, try catch your subprocess.run and print out error codes, exceptions etc.

File: bin/test-parse (hunk 4 above)
Comment (kik-kik) [reply bin/test-parse:93]: Ok, I'd say I'll go with the SQL approach. As Jeff mentioned this way we are already interacting with parsed values by Airflow so we can be sure they are set correctly.

For the above reason, I will leave the script execution where it correctly is in order to avoid having to spin up Airflow locally twice in a row (docker-compose up) as it will just extend the time the pipeline takes to complete.

File: bin/test-dag-tags.py (hunk 1 above)
Comment (kik-kik) [reply bin/test-dag-tags.py:16]: So in this case I made a small tweak to the `.run` call, I believe there is no need for any additional error handling here for that specific command and output should be redirected to `STDOUT` and exception should be raised if `non-zero` return code is present for the shell command.

Comment (kik-kik) [reply bin/test-dag-tags.py]: apologies, it looks like I actually have to manually output the errors. Adding that now.

**Comment by kik-kik** [comment]
@haroldwoo @jklukas ok, I'd say this is the final version. Please let me know if you have any comments.

File: bin/test-dag-tags.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    return _execute_shell_cmd("docker-compose exec web mysql -Ns -h db -u root -p{pswd} airflow -e 'SELECT dag_id from dag WHERE is_subdag = 0;'", cmd_params)
+
+
+if __name__ == "__main__":
+    db_pass = "secret"
+    # Assumes the web and db containers are already running
+    tag_errors = 0
+
+    dags = get_number_of_dags_loaded(db_pass)
```
Comment (haroldwoo) [inline bin/test-dag-tags.py:85]: nit: func name implies it returns length. change name to something like get_loaded_dags_from_sql

File: bin/test-dag-tags.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        key: val["value"]
+        for key, val in cmd_params.items()
+    }
+
+    cmd_output = subprocess.run(
+        cmd.format(**cmd_params_formatted),
+        shell=True,
+        capture_output=True,
+        text=True,
+    )
```
Comment (haroldwoo) [inline bin/test-dag-tags.py:35]: nit: whitespace

File: .circleci/config.yml [hunk 10]
```diff
@@ -58,8 +58,9 @@ jobs:
             docker-compose build
             # now take ownership of the folder
             sudo chown -R 10001:10001 .
+
       - run:
-          name: Test if dag scripts can be parsed
+          name: Test if dag scripts can be parsed by Airflow and that tags have been correctly set
```
Comment (haroldwoo) [inline .circleci/config.yml:63]: maybe add a comment noting that valid tags are defined in  bin/test-dag-tags.py

File: bin/test-dag-tags.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        print(cmd_output.stdout)
+        print(cmd_output.stderr)
+        raise _err
+
+    print("Command executed successfully, processing output...")
+
+    return cmd_output.stdout.strip().replace("\r", "").split("\n")
+
+
+def retrieve_existing_airflow_dags_from_db(pswd: str) -> Dict[str, str]:
```
Comment (haroldwoo) [inline bin/test-dag-tags.py:48]: nit: name get_dags_and_tags_from_sql or similar to whatever you rename the func on line 70 to

**Review by haroldwoo — APPROVED** [review]
lgtm, just had some nits

File: bin/test-dag-tags.py (hunk 8 above)
Comment (kik-kik) [reply bin/test-dag-tags.py:85]: that's a fair point, I believe originally that's what it did then I changed it and forgot to update the fun name. Fixed that now.

File: bin/test-dag-tags.py (hunk 9 above)
Comment (kik-kik) [reply bin/test-dag-tags.py:35]: Not sure I got it, you mean add a newline between `cmd_output` and `try`?
